libbpf-cargo/src/lib.rs

```diff
@@ -240,15 +262,25 @@ impl SkeletonBuilder {
             )));
         }
 
-        if self.obj.is_none() {
-            let name = filename.split('.').next().unwrap();
+        if self.obj.is_none() || self.extract_vmlinux_header {
             let dir = tempdir().map_err(|e| Error::Build(e.into()))?;
-            let objfile = dir.path().join(format!("{name}.o"));
-            self.obj = Some(objfile);
             // Hold onto tempdir so that it doesn't get deleted early
             self.dir = Some(dir);
         }
 
+        if self.obj.is_none() {
+            let name = filename.split('.').next().unwrap();
+            // Unwrap is safe here since we guarantee that dir.is_some() above
+            let objfile = self.dir.as_ref().unwrap().path().join(format!("{name}.o"));
```

Comment:
Let's use `SANITY` style comment, similar to here:
https://github.com/libbpf/libbpf-rs/blob/855eb8422e98e4632d433d5fbf065ab703674fbf/libbpf-rs/src/query.rs#L291-L292

```suggestion
            // SANITY: `dir` is guaranteed to be `Some` at this point.
            let objfile = self.dir.as_ref().unwrap().path().join(format!("{name}.o"));
```